addons/barcodes/models/barcode_rule.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+import re
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class BarcodeRule(models.Model):
+    _name = 'barcode.rule'
+    _order = 'sequence asc'
+
+
+    def _get_type_selection(self):
+        return [('alias','Alias'),('product','Unit Product')]
+
+    #indirection for inheritance
+    _type_selection = lambda self: self._get_type_selection()
+
+    name = fields.Char(string='Rule Name', required=True, help='An internal identification for this barcode nomenclature rule')
+    barcode_nomenclature_id = fields.Many2one('barcode.nomenclature', string='Barcode Nomenclature')
+    sequence = fields.Integer(string='Sequence', help='Used to order rules such that rules with a smaller sequence match first')
+    encoding = fields.Selection([
+        ('any', 'Any'),
+        ('ean13', 'EAN-13'),
+        ('ean8', 'EAN-8'),
+        ('upca', 'UPC-A'),
+    ], string='Encoding', required=True, default='any', help='This rule will apply only if the barcode is encoded with the specified encoding')
+    type = fields.Selection(_type_selection, string='Type', required=True, default='product')
+    pattern = fields.Char(string='Barcode Pattern', help="The barcode matching pattern", required=True, default='.*')
+    alias = fields.Char(string='Alias', default='0', help='The matched pattern will alias to this barcode', required=True)
+
+    @api.constrains('pattern')
+    def _check_pattern(self):
+        for record in self:
+            p = re.sub(r'\\\\|\\{|\\}', "X", record.pattern)
+            findall = re.findall("[{]|[}]", p) # p does not contain escaped { or }
+            if len(findall) == 2:
+                if not re.search("[{][N]*[D]*[}]", p):
+                    raise ValidationError(_("There is a syntax error in the barcode pattern ") + record.pattern + _(": braces can only contain N's followed by D's."))
+                elif re.search("[{][}]", p):
+                    raise ValidationError(_("There is a syntax error in the barcode pattern ") + record.pattern + _(": empty braces."))
+            elif len(findall) != 0:
+                raise ValidationError(_("There is a syntax error in the barcode pattern ") + record.pattern + _(": a rule can only contain one pair of braces."))
+            elif p == '*':
+                raise ValidationError(_(" '*' is not a valid Regex Barcode Pattern. Did you mean '.*' ?"))
```

Comment:
You don't think this method needs to optimize ???
![Uploading 10944_600.gif…]()

After optimization always code --- "ohh really i am looking like this ??" :laughing:  
take it own coffee.